components/eam/src/physics/cam/micro_p3.F90

```diff
@@ -2855,6 +2862,125 @@ subroutine prevent_ice_overdepletion(pres,t_atm,qv,latent_heat_sublim,inv_dt,
 
 end subroutine prevent_ice_overdepletion
 
+subroutine water_vapor_conservation(qv,qidep,qinuc,qi2qv_sublim_tend,qr2qv_evap_tend,dt)
+
+  implicit none
+
+  real(rtype), intent(in)     :: qv,qi2qv_sublim_tend,qr2qv_evap_tend,dt
+  real(rtype), intent(inout) :: qidep,qinuc
+  real(rtype)                :: qv_avail, qv_sink, ratio
+
+  qv_avail = qv + (qi2qv_sublim_tend+qr2qv_evap_tend)*dt
+  qv_sink  = (qidep + qinuc)*dt
+
+  if (qv_sink > qv_avail .and. qv_sink>1.e-20_rtype) then
+     ratio = qv_avail/qv_sink
+     qidep = qidep*ratio
+     qinuc = qinuc*ratio
+  endif
+
+  return
+end subroutine water_vapor_conservation
+
+subroutine ice_supersat_conservation(qidep,qinuc,cld_frac_i,qv,qv_sat_i,latent_heat_sublim,T_atm,dt)
+  !Make sure ice processes don't drag qv below ice supersaturation
+
+  implicit none
+
+  real(rtype), intent(in) :: cld_frac_i,qv,qv_sat_i,latent_heat_sublim,T_atm,dt
+  real(rtype), intent(inout) :: qidep,qinuc
+
+  real(rtype) :: qv_sink, qv_avail, fract
+
+  qv_sink = qidep + qinuc ! in [kg/kg] cell-avg values
+
+  if (qv_sink > qsmall .and. cld_frac_i > 1.0e-20_rtype) then
+     ! --- Available water vapor for deposition/nucleation
+     qv_avail = (qv - qv_sat_i) / &
+          (1.0_rtype + bfb_square(latent_heat_sublim)*qv_sat_i / (cp*rv*bfb_square(T_atm)) ) / dt
+
+     ! --- Only excess water vapor can be limited
+     qv_avail = max(qv_avail,0.0_rtype)
+
+     if (qv_sink >  qv_avail) then
+        fract = qv_avail / qv_sink
+        qinuc = qinuc * fract
+        qidep = qidep * fract
+     endif
+  endif
+
+  return
+end subroutine ice_supersat_conservation
+
+subroutine nc_conservation(nc, nc_selfcollect_tend, dt, nc_collect_tend, nc2ni_immers_freeze_tend, &
```

Comment:
For newcomers it would be exceedingly helpful to have a comment or two that describes each of these subroutines and their purpose in succinct language.

Reply:
done.